ui/src/state/notifications.rs

```diff
@@ -0,0 +1,148 @@
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    config::Configuration,
+    utils::{notifications::set_badge, sounds::Sounds},
+};
+
+// This kind is used to determine which notification kind to add to. It can also be used for querying specific notification counts.
+pub enum NotificaitonKind {
+    FriendRequest,
+    Message,
+    Settings,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct Notifications {
+    pub friends: u32, // For notifications about new friends, friend requests and related CTAs.
+    pub messages: u32, // For notifications about new messages, mentions.
+    pub settings: u32, // For notifications about updates, issues and more.
+}
+
+impl Notifications {
+    pub fn new() -> Self {
+        // By default we'll say there are no notifications.
+        Self {
+            // Represents the total notification count for all friend events.
+            friends: 0,
+            // Represents the total notification count for all message events. Including all conversations and groups.
+            messages: 0,
+            // Represents total notification count for all settings events. E.g. updates, issues, etc.
+            settings: 0,
+        }
+    }
+
+    // This method is used for calculating the badge count for the app tray icon.
+    pub fn total(&self) -> u32 {
+        // Since this is read only, we can just load the config here.
+        let config = Configuration::load_or_default();
+
+        let mut total = 0;
+
+        // Only count notifications that are enabled in the config.
+        if config.notifications.friends_notifications {
+            total += self.friends;
+        }
+        if config.notifications.messages_notifications {
+            total += self.messages;
+        }
+        if config.notifications.settings_notifications {
+            total += self.settings;
+        }
+
+        total
+    }
+
+    // Adds notification(s) to the specified kind.
+    pub fn increment(&mut self, kind: NotificaitonKind, count: u32) {
+        match kind {
+            NotificaitonKind::FriendRequest => self.friends += count,
+            NotificaitonKind::Message => self.messages += count,
+            NotificaitonKind::Settings => self.settings += count,
+        };
+
+        // Update the badge any time notifications are added.
+        let _ = set_badge(self.total());
+
+        // Plays the notification sound.
+        let config = Configuration::load_or_default();
+        if config.notifications.enabled {
+            crate::utils::sounds::Play(Sounds::Notification);
+        }
+    }
+
+    // Removes notification(s) from the specified kind.
+    pub fn decrement(&mut self, kind: NotificaitonKind, count: u32) {
+        match kind {
+            NotificaitonKind::FriendRequest => {
+                // Prevent underflow.
+                if count > self.friends {
```

Comment:
this is fine. if you really want to be fancy, you can make it a one liner using [this](https://doc.rust-lang.org/std/primitive.u32.html#method.checked_sub)

`self.friends = self.friends.checked_sub(count).unwrap_or_default(); `